front-end-tools/src/Main.tsx

```diff
@@ -1081,19 +1445,974 @@ export default function Main(props: ConnectionProps) {
+                                                                    setReadError((err as Error).message);
+                                                                });
+                                                        })
+                                                        .catch((err: Error) => setReadError((err as Error).message));
+                                                }
+                                            }}
+                                        />
+                                    </Form.Group>
+                                </div>
+
+                                {!attributeForm.watch('deriveFromSmartContractIndex') && (
+                                    <Form.Group className="mb-3">
+                                        <Form.Label>
+                                            Upload Smart Contract Module Schema File (e.g. schema.bin):
+                                        </Form.Label>
+                                        <Form.Control
+                                            type="file"
+                                            accept=".bin"
+                                            {...attributeForm.register('file')}
+                                            onChange={async (e) => {
+                                                const fileRegister = attributeForm.register('file');
+
+                                                fileRegister.onChange(e);
+
+                                                setUploadErrorRead(undefined);
+                                                setUploadedModuleSchemaBase64Read(undefined);
+
+                                                const files = attributeForm.getValues('file');
+
+                                                if (files !== undefined && files !== null && files.length > 0) {
+                                                    const file = files[0];
+                                                    const arrayBuffer = await file.arrayBuffer();
+
+                                                    const schema = btoa(
+                                                        new Uint8Array(arrayBuffer).reduce((data, byte) => {
+                                                            return data + String.fromCharCode(byte);
+                                                        }, '')
+                                                    );
+                                                    setUploadedModuleSchemaBase64Read(schema);
+                                                    console.log(schema);
+                                                } else {
+                                                    setUploadErrorRead('Upload schema file is undefined');
+                                                }
+                                            }}
+                                        />
```

Comment:
Here also it would be more suitable with another solution in my opinion:

```tsx
function validateResult(input: string | Error): true | string { // returns validation message if invalid
    ...
}
...
<Controller
    name="file"
    rules={{ 
        required:"Schema file is required",
        validate: validateResult // Validates whether an error/success result was produced (maybe not needed)?
    }}
    render={(field) =>
        <Form.Control
            type="file"
            accept=".bin"
            {...field}
            onChange={async (e) => {
                const files = e.target.value; // or whatever works
                                                
                if (files === null || files?.length === 0) {
                    field.onChange(undefined); // will trigger "required" validation rule
                    return;
                }
                                                
                try {
                     const file = files[0];
                     const arrayBuffer = await file.arrayBuffer();

                     const schema = btoa(
                         new Uint8Array(arrayBuffer).reduce((data, byte) => {
                             return data + String.fromCharCode(byte);
                         }, '')
                     );
                                                
                     field.onChange(schema);
                 } catch (e) {
                     field.onChange(e); // If it'll never fail, the try/catch and the associated "validate" rule can be removed
                 }
             }}
         />
     }
/>
```

Reply:
Actually.. I don't know the best way to do this. The problem is, that you need the value of the original field (the inner field), to stay the same, while deriving something from the value whenever it changes. I haven't tested this, but what I think I would do would be:

```tsx
type DerivedInput<O, D> = {
    original: O;
    derived: D;
};

function validateFileInput(input: DerivedInput<FileInput, Base64String | Error>): true | string {
    ...
}
...
<Controller
    name="file"
    rules={{ 
        validate: validateFileInput // Validates whether an error/success result was produced (maybe not needed)?
    }}
    render={(field) =>
        <Form.Control
            type="file"
            accept=".bin"
            {...field}
            value={field.value.original}
            onChange={async (e) => {
                const files = e.target.value; // or whatever works
                                                
                if (files === null || files?.length === 0) {
                    field.onChange({original: files, derived: new Error('No file found in input')});
                    return;
                }
                                                
                try {
                     const file = files[0];
                     const arrayBuffer = await file.arrayBuffer();

                     const schema = btoa(
                         new Uint8Array(arrayBuffer).reduce((data, byte) => {
                             return data + String.fromCharCode(byte);
                         }, '')
                     );
                                                
                     field.onChange({original: files, derived: schema});
                 } catch (e) {
                     field.onChange({original: files, derived: e}); // If it'll never fail, the try/catch and the associated "validate" rule can be removed
                 }
             }}
         />
     }
/>
```

Sorry if it's all a bit confusing.. I actually don't know how to do this the best way, would probably need to find out through trial and error 🤷